cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,525 @@
+            //     // rethink: sort the id .
+            tileNo = _tile_table_info[0];
+            _tile_table.insert( tileKey, tileNo );
+            //         Kokkos::atomic_fetch_add( &( _tile_table_info[0] ),
+            //         (ValueType)1
+            //         );
+        }
+        // else
+        // {
+        //     tileNo = _tile_table.value_at( _tile_table.find( tileKey ) );
+        // }
+        return tileNo;
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    void sortTile()
+    {
+        // pass
+    }
+
+    // query
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType find( int tile_i, int tile_j, int tile_k )
+    {
+        return _tile_table.value_at(
+            _tile_table.find( _op_ijk2key( tile_i, tile_j, tile_k ) ) );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    KeyType ijk_to_key( int tile_i, int tile_j, int tile_k )
+    {
+        return _op_ijk2key( tile_i, tile_j, tile_k );
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    void key_to_ijk( KeyType &key, int &tile_i, int &tile_j, int &tile_k )
+    {
+        return _op_key2ijk( key, tile_i, tile_j, tile_k );
+    }
+
+  private:
+    //! [0] current valid table size, [1] pre-allocated size
+    Kokkos::Array<ValueType, 2> _tile_table_info;
+    //! default factor, rehash by which when need more capacity of the hash
+    //! table
+    Kokkos::Array<float, 1> _rehash_coeff;
+    //! hash table (tile IJK <=> tile No)
+    Kokkos::UnorderedMap<KeyType, ValueType, ExecutionSpace> _tile_table;
+    //! Op: tile IJK <=> tile No
+    TileID2HashKey<KeyType, HashType> _op_ijk2key;
+    HashKey2TileID<KeyType, HashType> _op_key2ijk;
+};
+
+//---------------------------------------------------------------------------//
+template <int CBits, int CNumPerDim, int CNumPerTile>
+class TileIndexSpace
+{
+  public:
+    //! Number of cells inside each tile
+    static constexpr int CellBitsPerTileDim = CBits;
+    static constexpr int CellNumPerTileDim = CNumPerDim;
+    static constexpr int CellNumPerTile = CNumPerTile;
+    static constexpr int Rank = 3;
+
+    //! Cell ijk <=> Cell Id
+    template <typename... Coords>
+    KOKKOS_FORCEINLINE_FUNCTION static constexpr auto
+    coord_to_offset( Coords &&... coords ) -> uint64_t
+    {
+        return from_coord<Coord2OffsetDim>( std::forward<Coords>( coords )... );
+    }
+
+    template <typename Key, typename... Coords>
+    KOKKOS_FORCEINLINE_FUNCTION static constexpr void
+    offset_to_coord( Key &&key, Coords &... coords )
+    {
+        to_coord<Offset2CoordDim>( std::forward<Key>( key ), coords... );
+    }
+
+  protected:
+    //! Coord  <=> Offset Computations
+    struct Coord2OffsetDim
+    {
+        template <typename Coord>
+        KOKKOS_FORCEINLINE_FUNCTION constexpr auto operator()( int dimNo,
+                                                               Coord &&i )
+            -> uint64_t
+        {
+            uint64_t result = (uint64_t)i;
+            for ( int i = 0; i < dimNo; i++ )
```

Comment:
Collision between name of the function parameter and the loop counter variable is unfortunate.  Consider renaming one of the two.